kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java

```diff
@@ -75,32 +80,154 @@ public static Commit getActiveCommitAtTimestamp(
+   * already proven that `searchTimestamp` is in the given range.
+   */
+  private static Commit getActiveCommitAtTimeFromICTRange(
+      long searchTimestamp,
+      long startCommitVersionInclusive,
+      long endCommitVersionInclusive,
+      Engine engine,
+      Path logPath)
+      throws IOException {
+    // Now we have a range of commits to search through. We can use binary search to find the
+    // commit that is closest to the search timestamp.
+    Optional<Tuple2<Long, Long>> greatestLowerBoundOpt =
+        InCommitTimestampUtils.greatestLowerBound(
+            searchTimestamp,
+            startCommitVersionInclusive,
+            endCommitVersionInclusive,
+            version -> CommitInfo.getRequiredInCommitTimestampFromFile(engine, logPath, version));
+    // This indicates that the search timestamp is less than the earliest commit.
+    if (!greatestLowerBoundOpt.isPresent()) {
+      long startIct =
+          CommitInfo.getRequiredInCommitTimestampFromFile(
+              engine, logPath, startCommitVersionInclusive);
+      return new Commit(startCommitVersionInclusive, startIct);
+    }
+    Tuple2<Long, Long> greatestLowerBound = greatestLowerBoundOpt.get();
+    return new Commit(greatestLowerBound._1, greatestLowerBound._2);
+  }
+
+  /**
+   * Gets the commit that enabled in-commit timestamps.
+   *
+   * @param snapshot The latest snapshot of the table. This is used to determine when in-commit
+   *     timestamps were enabled.
+   * @param earliestCommit The earliest commit under consideration. If in-commit timestamps were
+   *     enabled for the entire history, this function will return this commit.
+   * @return The commit that enabled in-commit timestamps. If the table does not have in-commit
+   *     timestamps enabled, this will be the commit after the latest version. If in-commit
+   *     timestamps were enabled for the entire history, this will be `earliestCommit`.
+   */
+  private static Commit getICTEnablementCommit(SnapshotImpl snapshot, Commit earliestCommit) {
+    Metadata metadata = snapshot.getMetadata();
+    if (!IN_COMMIT_TIMESTAMPS_ENABLED.fromMetadata(metadata)) {
+      // Pretend ICT will be enabled after the latest version and requested timestamp.
+      // This will force us to use the non-ICT search path.
+      return new Commit(snapshot.getVersion() + 1, Long.MAX_VALUE);
+    }
+    Optional<Long> enablementTimestampOpt =
+        IN_COMMIT_TIMESTAMP_ENABLEMENT_TIMESTAMP.fromMetadata(metadata);
+    Optional<Long> enablementVersionOpt =
+        IN_COMMIT_TIMESTAMP_ENABLEMENT_VERSION.fromMetadata(metadata);
+    if (enablementTimestampOpt.isPresent() && enablementVersionOpt.isPresent()) {
+      return new Commit(enablementVersionOpt.get(), enablementTimestampOpt.get());
+    } else if (!enablementTimestampOpt.isPresent() && !enablementVersionOpt.isPresent()) {
+      // This means that ICT has been enabled for the entire history.
```

Comment:
why is this the case? If my table doesn't support ICT --> then won't these fields be empty?